elasticapm/utils/disttracing.py

```diff
@@ -119,13 +122,86 @@ def merge_duplicate_headers(cls, headers, key):
-                if item[0] == key
-            ])
+            return ",".join([item[1] for item in headers if item[0] == key])
         return headers.get(key)
 
+    def _parse_tracestate(self, tracestate):
+        """
+        Tracestate can contain data from any vendor, made distinct by vendor
+        keys. Vendors are comma-separated. The elastic (es) tracestate data is
+        made up of key:value pairs, separated by semicolons. It is meant to
+        be parsed into a dict.
+
+            tracestate: es=key:value;key:value...,othervendor=<opaque>
+        """
+        if not tracestate:
+            return {}
+        if "es=" not in tracestate:
+            return {}
+
+        ret = {}
+        state = re.match(r"es=([^,]*)", tracestate).group(1).split(";")
+        for keyval in state:
+            if not keyval:
+                continue
+            key, _, val = keyval.partition(":")
+            ret[key] = val
+
+        return ret
+
+    def _set_tracestate(self):
+        elastic_value = ";".join(["{}:{}".format(k, v) for k, v in compat.iteritems(self.tracestate_dict)])
+        # No character validation needed, as we validate in `add_tracestate`. Just validate length.
+        if len(elastic_value) > 256:
+            logger.debug("Modifications to TraceState would violate length limits, ignoring.")
+            raise TraceStateFormatException()
+        elastic_state = "es={}".format(elastic_value)
+        if not self.tracestate:
+            return elastic_state
+        else:
+            # Remove es=<stuff> from the tracestate, and add the new es state to the end
+            otherstate = re.sub(r"es=([^,]*),?", "", self.tracestate)
+            # No validation of `otherstate` required, since we're downstream. We only need to check `es=`
+            # since we introduced it, and that validation has already been done at this point.
+            if otherstate:
+                if otherstate[-1] == ",":
+                    return "{}{}".format(otherstate, elastic_state)
+                else:
+                    return "{},{}".format(otherstate, elastic_state)
+            else:
+                return elastic_state
+
+    def add_tracestate(self, key, val):
+        """
+        Add key/value pair to the tracestate.
+
+        We do most of the validation for valid characters here. We have to make
+        sure none of the reserved separators for tracestate are used in our
+        key/value pairs, and we also need to check that all characters are
+        within the valid range. Checking here means we never have to re-check
+        a pair once set, which saves time in the _set_tracestate() function.
+        """
+        key = str(key)
+        val = str(val)
+        for bad in (":", ";", ",", "="):
+            if bad in key or bad in val:
+                logger.debug("New tracestate key/val pair contains invalid character '{}', ignoring.".format(bad))
+                return
+        for c in key + val:
```

Comment:
depending on how large `val` can become, it might be worth avoiding creating a new string just to throw it away again. Something like

```suggestion
        for c in itertools.chain(key, val)
```

(requires an `import itertools` up top of course)

Reply:
`val` should always be <256 characters, but since we're not checking it at this point, but rather at the tracestate string generation, I'll use your suggestion. Thanks.